### PR TITLE
earthly ls: lists targets under an Earthfile

### PR DESCRIPTION
**Pull request by alexcb**
Displays a list of all targets under an Earthfile.

usage:

    earthly ls

    earthly ls ./sub-dir-containing-an-earthfile

This commit is dedicated to jazzdan.

Signed-off-by: Alex Couture-Beil <alex@earthly.dev>

### CODE REVIEW

**Comment by alexcb** [comment]
example:
```
$ earthly ls 
all
all-buildkitd
all-dind
base
changelog
code
debugger
deps
dind
dind-alpine
dind-ubuntu
earthly
earthly-all
earthly-darwin-amd64
earthly-darwin-arm64
earthly-docker
earthly-integration-test-base
earthly-linux-amd64
earthly-linux-arm64
earthly-linux-arm7
earthly-windows-amd64
examples
examples1
examples2
for-darwin
for-darwin-m1
for-linux
for-own
for-windows
lint
lint-changelog
lint-newline-ending
lint-scripts
lint-scripts-auth-test
lint-scripts-base
lint-scripts-misc
markdown-spellcheck
prerelease
prerelease-script
shellrepeater
test
test-all
unit-test
update-buildkit
vale
```

and a sub dir:

```
$ earthly ls ./buildkitd
base
buildkit-sha
buildkitd
update-buildkit
```

File: cmd/earthly/main.go [hunk 1]
```diff
@@ -704,6 +705,12 @@ func newEarthlyApp(ctx context.Context, console conslogging.ConsoleLogger) *eart
 				},
 			},
 		},
+		{
+			Name:      "ls",
+			Usage:     "List targets from an earthfile *experimental*",
```
Comment (andrewsykim) [inline cmd/earthly/main.go:710]: nit: s/earthfile/Earthfile/

File: tests/Earthfile [hunk 2]
```diff
@@ -901,6 +903,18 @@ help:
...
 
+ls:
+    COPY ls.earth Earthfile
+    RUN earthly ls 2>/dev/null | tee actual
+    RUN echo -e "alpha\nbase\nbravo\ncharlie" > expected
+    RUN diff expected actual
+
+ls-subdir:
+    COPY ls.earth subdir/Earthfile
+    RUN earthly ls ./subdir 2>/dev/null | tee actual
```
Comment (andrewsykim) [inline tests/Earthfile:914]: Should this also work if the user used the Earthfile path too? 
```
earthly ls ./subdir/Earthfile
```

File: cmd/earthly/main.go [hunk 3]
```diff
@@ -704,6 +705,12 @@ func newEarthlyApp(ctx context.Context, console conslogging.ConsoleLogger) *eart
...
+			UsageText: "earthly [options] ls [<import-path>]",
```
Comment (andrewsykim) [inline cmd/earthly/main.go:711]: Does "import-path" here mean the path can also be the path to the Earthfile, like `earthly ls ./subdir/Earthfile`? Or is it strictly path to the directory where the Earthfile is?

Wondering if the latter should even work -- I can see it being useful but maybe not worth the extra complexity here?

**Review by andrewsykim — COMMENTED** [review]
Just some questions

File: tests/Earthfile [hunk 4]
```diff
@@ -901,6 +903,18 @@ help:
     ENV EARTHLY_SHOW_HIDDEN=1
     RUN earthly --help | grep 'buildkit-volume-name'
 
+ls:
+    COPY ls.earth Earthfile
+    RUN earthly ls 2>/dev/null | tee actual
```
Comment (adamgordonbell) [inline tests/Earthfile:908]: Testing earthly with earthly never ceases to impress me

**Review by adamgordonbell — APPROVED** [review]
🎉

File: cmd/earthly/main.go (hunk 3 above)
Comment (alexcb) [reply cmd/earthly/main.go:711]: I took a closer look at our docs, and under https://docs.earthly.dev/docs/guides/target-ref we call this `<project-ref>`. I'll rename import-path to project-ref.

I wanted to keep it consistent with the rest of our target referencing (where we never type out the word "Earthfile" on our targets). We could potentially trim any `/Earthfile` suffix if it's passed in by our users. But then that causes us to diverge from what a project-reference is.

Comment (alexcb) [reply cmd/earthly/main.go]: thinking it over a bit more, it's a pretty easy thing to do, let's see if we get a follow up bug report for "unable to list targets from `Earthfile` contained in sub-directory named `Earthfile` (i.e. `earthly ls ./Earthfile/Earthfile`).

**Comment by vladaionescu** [comment]
Wonder if the targets should have `+` in front of them, for consistency with output in other places.

Another thought.. I wonder if this should be connected somehow to the earthlydoc feature https://github.com/earthly/earthly/issues/583?

e.g. `earthly doc ls`? (and we implement the rest later...)

Perhaps we can talk more about some subtle usability choices during standup tomorrow.

**Comment by vladaionescu** [comment]
Wonder if it's easy to grab the `ARG`s too from the AST and list their names (and maybe default values)?

Also, don't forget about release notes 😉

**Comment by alexcb** [comment]
> Wonder if the targets should have `+` in front of them, for consistency with output in other places.

I was tempted to include that.

And if we allow for `earthly ls -l`, we could display the full `<project-ref>+` prefix in front, e.g.

```
$ earthly ls -l ./buildkitd
./buildkitd+base
./buildkitd+buildkit-sha
./buildkitd+buildkitd
./buildkitd+update-buildkit
```

> 
> Another thought.. I wonder if this should be connected somehow to the earthlydoc feature #583?

definitely a possibility -- maybe a flag could toggle it on/off?

**Comment by alexcb** [comment]
> Wonder if it's easy to grab the `ARG`s too from the AST and list their names 

earthly ls -la: show args and project-ref
Running without flags will display:

    alex@mah:~/gh/earthly/earthly$ ./build/linux/amd64/earthly ls ./buildkitd
    +base
    +buildkit-sha
    +buildkitd
    +update-buildkit

Running with -l will display the project-ref:

    alex@mah:~/gh/earthly/earthly$ ./build/linux/amd64/earthly ls -l ./buildkitd
    ./buildkitd+base
    ./buildkitd+buildkit-sha
    ./buildkitd+buildkitd
    ./buildkitd+update-buildkit

Running with -a will display the arguments for each target:

    alex@mah:~/gh/earthly/earthly$ ./build/linux/amd64/earthly ls -la ./buildkitd
    ./buildkitd+base
    ./buildkitd+buildkit-sha
      --BUILDKIT_GIT_SHA
      --BUILDKIT_GIT_BRANCH
    ./buildkitd+buildkitd
      --BUILDKIT_PROJECT
      --EARTHLY_GIT_HASH
      --EARTHLY_TARGET_TAG_DOCKER
      --TAG
      --DOCKERHUB_USER
    ./buildkitd+update-buildkit
      --BUILDKIT_GIT_SHA

> (and maybe default values)?

I think that's best left for another PR, as that's not provided by the auto-completion helpers that this PR is making use of.

> 
> Also, don't forget about release notes wink

hah, thanks for the reminder :)

**Comment by vladaionescu** [comment]
> definitely a possibility -- maybe a flag could toggle it on/off?

I don't think a flag is necessary, given that this is a feature that the user anyway chooses to use every time.

Maybe `earthly ls` could become in the future the `earthly doc` feature itself. `ls` does have a certain familiarity & simplicity to it, which is nice.

**Comment by vladaionescu** [comment]
The `-l` and `-la` variants look really nice!

**Review by vladaionescu — APPROVED** [review]

File: cmd/earthly/main.go [hunk 5]
```diff
@@ -3002,6 +3025,68 @@ func ifNilBoolDefault(ptr *bool, defaultValue bool) bool {
...
+	if targetToParse == "" {
+		targetToDisplay = "current directory"
+	}
+
+	gitLookup := buildcontext.NewGitLookup(app.console, app.sshAuthSock)
+	resolver := buildcontext.NewResolver("", nil, gitLookup, app.console, "")
+	var gwClient gwclient.Client // TODO this is a nil pointer which causes a panic if we try to expand a remotely referenced earthfile
+	// it's expensive to create this gwclient, so we need to implement a lazy eval which returns it when required.
+
+	target, err := domain.ParseTarget(fmt.Sprintf("%s+base", targetToParse)) //the +base is required to make ParseTarget work; however is ignored by GetTargets
```
Comment (brandonSc) [inline cmd/earthly/main.go:3054]: super minor: `//the` vs typical style in the project `// the`

**Pull request merged**